### ISSUE TITLE
Top-k Voting-nms and box_voting

### ISSUE THREAD

**Issue by sbl1996** [PaddlePaddle/PaddleDetection#452]
Is the implemented `box_voting` equal to top-k voting-nms as described in [2nd Place Solution in Google AI Open Images Object Detection Track 2019](https://arxiv.org/abs/1911.07171)?

If not, would you tell me the difference between them? Thanks very much!

https://github.com/PaddlePaddle/PaddleDetection/blob/f605edb9c97613ee2730260143c0aa128710f300/ppdet/utils/post_process.py#L116-L128

**Comment by littletomatodonkey** [comment]
Hi, thanks for your attention. There are some differences.
1. For boxes in the same position(chosen based on vote_thresh), box voting in the repo uses all the boxes, but the top-k voting method just chooses `top-k` bboxes. And during merging bboxes, bbox voting just uses the highest score for current bbox group, but top-k voting method averages scores for  all the bboxes in the same group.
2. Box voting can merge the nms results, but not verified on softnms results, Top-k voting method is verified for softnms results on Open Images Dataset(OID).(Actually we tried using box-voting for OID softnms results, but found it just has a limited improvement)
3. Since top-k voting method is used to merge softnms results, the hyper-params, `vote_thresh` and `score_thres` need also to be finetuned on specified dataset.

**Comment by sbl1996** [comment]
This line `top_dets[k, 1:] = np.average(boxes_to_vote, axis=0, weights=ws)` shows that box_voting merges boxes by averaging all boxes in the same group using scores as weights rather than only selecting the one with the highest score.

**Comment by littletomatodonkey** [comment]
> This line `top_dets[k, 1:] = np.average(boxes_to_vote, axis=0, weights=ws)` shows that box_voting merges boxes by averaging all boxes in the same group using scores as weights rather than only selecting the one with the highest score.

Thanks for correcting that, I have deleted the error~

**Comment by littletomatodonkey** [comment]
> 2. Does top-k voting use simple average or weighted average?
For bboxes, we use weighted average based on the normalized scores(all the scores are normalized for the sum to be 1).
For scores, we use weighted averge based on different methods' metric.

**Comment by sbl1996** [comment]
Is k equal to or close to the number of models for ensemble? Or you just tune it using a validation set?

**Comment by littletomatodonkey** [comment]
> Is k equal to or close to the number of models for ensemble? Or you just tune it using a validation set?

Actually we find when k is close to the number of models, the final results tends to perform better(The conclusion is got from validation set, and we found it is also applicable for the final submission).
